influxdb3_cache/src/meta_cache/mod.rs

```diff
@@ -0,0 +1,854 @@
+                })
+                .unwrap_or_default(),
+            PredicateKind::NotEq(rhs) => self
+                .0
+                .iter()
+                .filter(|(v, (t, _))| t > &expired_time_ns && *v != rhs)
+                .map(|(v, (_, n))| (v.clone(), n.as_ref()))
+                .collect(),
+            PredicateKind::In(in_list) => in_list
+                .iter()
+                .filter_map(|v| {
+                    self.0.get_key_value(v).and_then(|(v, (t, n))| {
+                        (t > &expired_time_ns).then(|| (v.clone(), n.as_ref()))
+                    })
+                })
+                .collect(),
+            PredicateKind::NotIn(not_in_set) => self
+                .0
+                .iter()
+                .filter(|(v, (t, _))| t > &expired_time_ns && !not_in_set.contains(v))
+                .map(|(v, (_, n))| (v.clone(), n.as_ref()))
+                .collect(),
+        }
+    }
+}
+
+/// A cache value, which for now, only holds strings
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+struct Value(Arc<str>);
+
+impl From<&FieldData> for Value {
+    fn from(field: &FieldData) -> Self {
+        match field {
+            FieldData::Key(s) => Self(Arc::from(s.as_str())),
+            FieldData::Tag(s) => Self(Arc::from(s.as_str())),
+            FieldData::String(s) => Self(Arc::from(s.as_str())),
+            FieldData::Timestamp(_)
+            | FieldData::Integer(_)
+            | FieldData::UInteger(_)
+            | FieldData::Float(_)
+            | FieldData::Boolean(_) => panic!("unexpected field type for metadata cache"),
+        }
+    }
+}
+
+/// A predicate that can be applied when gathering [`RecordBatch`]es from a [`MetaCache`]
+///
+/// This is intended to be derived from a set of filter expressions in Datafusion
+#[derive(Debug)]
+pub struct Predicate {
+    column_id: ColumnId,
+    kind: PredicateKind,
+}
+
+#[cfg(test)]
+impl Predicate {
+    fn new_eq(column_id: ColumnId, rhs: impl Into<Arc<str>>) -> Self {
+        Self {
+            column_id,
+            kind: PredicateKind::Eq(Value(rhs.into())),
+        }
+    }
+
+    fn new_not_eq(column_id: ColumnId, rhs: impl Into<Arc<str>>) -> Self {
+        Self {
+            column_id,
+            kind: PredicateKind::NotEq(Value(rhs.into())),
+        }
+    }
+
+    fn new_in(column_id: ColumnId, in_vals: impl IntoIterator<Item: Into<Arc<str>>>) -> Self {
+        Self {
+            column_id,
+            kind: PredicateKind::In(in_vals.into_iter().map(Into::into).map(Value).collect()),
+        }
+    }
+
+    fn new_not_in(column_id: ColumnId, in_vals: impl IntoIterator<Item: Into<Arc<str>>>) -> Self {
+        Self {
+            column_id,
+            kind: PredicateKind::NotIn(in_vals.into_iter().map(Into::into).map(Value).collect()),
+        }
+    }
+}
+
+// TODO: remove this when fully implemented, for now just suppressing compiler warnings
```

Comment:
This is for a follow up PR when the metadata cache actually gets hooked up right?

Reply:
Yeah, the type will get used when I implement the DataFusion trait to hook this all up to the query path in https://github.com/influxdata/influxdb/issues/25544